libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1100,7 +1105,7 @@ void SMTEncoder::endVisit(Literal const& _literal)
 
 		addArrayLiteralAssertions(
 			*symbArray,
-			applyMap(_literal.value(), [&](auto const& c) { return smtutil::Expression{size_t(c)}; })
+			applyMap(_literal.value(), [&](auto const& c) { return smtutil::Expression{size_t(static_cast<unsigned char const&>(c))}; })
```

Comment:
Was this also buggy?

Reply:
Yes, if you check the example `string_literal_to_dynamic_bytes.sol ` that I have added with `develop`, the second assert fails, but it should hold (at least on my MacOs).
The problem is, from what I understood, that the implementation of `char` can be signed (it is implementation defined). On my local, it was interpreting byte `0xFF` as `-1` and not `255`. When you cast that to `size_t` you get `2^64 - 1` instead of the expected `255`.

Reply:
Ok!

Reply:
Actually, what if you cast directly to `size_t`?

Reply:
Not sure what you mean. The original code was casting `char` to `size_t` (C-style cast).
I tried replacing `unsigned char` in my `static_cast` with `size_t` and I got the buggy behaviour again.

Reply:
Ok. I find the whole thing kinda weird though

Reply:
What exactly? The original code or my change?
We could simplify the lambda to `[](char c) { return smtutil::Expression{size_t(static_cast<unsigned char>(c))}; }`
The references don't make much sense there. (And we also don't need `[&]`, `[]` is enough here, we are not capturing anything.)

It looks like we can even do `[](unsigned char c) { return smtutil::Expression{size_t(c)}; }` which is more readable.

Reply:
Ah yes, that's much better indeed.